creme/metrics/_confusion_matrix.pyx

```diff
@@ -0,0 +1,232 @@
+
+    This structure is used to keep updated statistics about a single-output classifier's
+    performance and to compute multiple evaluation metrics.
+
+    """
+    _fmt = '0.0f'
+
+    def __init__(self, classes=None):
+        self._init_classes = set(classes) if classes is not None else set()
+        self.classes = self._init_classes
+        self.sum_diag = 0.0
+        self.sum_row = defaultdict(float)
+        self.sum_col = defaultdict(float)
+        self.data = defaultdict(lambda: defaultdict(float))
+
+    def __setitem__(self, entry, double sample_weight):
+        if len(entry) != 2:
+            raise KeyError(f'Expected (true_idx, pred_idx) entry, received: {entry}')
+        y_true, y_pred = entry
+        self.update(y_true, y_pred, sample_weight)
+
+    def __getitem__(self, key):
+        """Syntactic sugar for accessing the counts directly."""
+        return self.data[key]
+
+    def update(self, y_true, y_pred, sample_weight=1.):
+        if sample_weight is None:
+            # Since we ca not set a default value in the signature
+            sample_weight = 1.0
+
+        self.classes.update([y_true, y_pred])
+        self.data[y_true][y_pred] += sample_weight
+
+        if y_true == y_pred:
+            self.sum_diag += sample_weight
+        self.sum_row[y_true] += sample_weight
+        self.sum_col[y_pred] += sample_weight
+        return self
+
+    def revert(self, y_true, y_pred, sample_weight=1.):
+        # Revert is equal to subtracting so we pass the negative sample_weight
+        self.update(y_true, y_pred, -sample_weight)
+        return self
+
+    @property
+    def n_classes(self):
+        return len(self.classes)
+
+    @property
+    def shape(self):
+        return self.n_classes, self.n_classes
+
+    def reset(self):
+        self.__init__(classes=self._init_classes)
+
+    def __repr__(self):
+        # The classes are sorted alphabetically for reproducibility reasons
+        classes = sorted(self.classes)
+
+        if not classes:
+            return ''
+
+        # Determine the required width of each column in the table
+        largest_label_len = max(len(str(c)) for c in classes)
+        largest_number_len = len(str(max(max(v for v in c.values()) for c in self.data.values())))
+        width = max(largest_label_len, largest_number_len) + 2
+
+        # Make a template to print out rows one by one
+        row_format = '{:>{width}}' * (len(classes) + 1)
+
+        # Write down the header
+        table = row_format.format('', *map(str, classes), width=width) + '\n'
+
+        # Write down the true labels row by row
+        table += '\n'.join((
+            row_format.format(
+                str(y_true),
+                *[f'{self.data[y_true][y_pred]:{self._fmt}}' for y_pred in classes],
+                width=width
+            )
+            for y_true in classes
+        ))
+
+        return table
+
+
+cdef class MultiLabelConfusionMatrix:
```

Comment:
Just thinking out loud, isn't `MultiLabelConfusionMatrix` essentially a `dict` of `ConfusionMatrix`?

Reply:
It is possible to see it that way. However, many operations can be easily performed by using a 3D array:

```python
sum_tp = data[:,1,1]
```